runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/PortablePipelineJarUtils.java

```diff
@@ -51,4 +83,102 @@
+      JsonFormat.parser().merge(contents, builder);
+    }
+  }
+
+  public static Pipeline getPipelineFromClasspath() throws IOException {
+    Pipeline.Builder builder = Pipeline.newBuilder();
+    parseJsonResource("/" + PIPELINE_PATH, builder);
+    return builder.build();
+  }
+
+  public static Struct getPipelineOptionsFromClasspath() throws IOException {
+    Struct.Builder builder = Struct.newBuilder();
+    parseJsonResource("/" + PIPELINE_OPTIONS_PATH, builder);
+    return builder.build();
+  }
+
+  public static ProxyManifest getArtifactManifestFromClassPath() throws IOException {
+    ProxyManifest.Builder builder = ProxyManifest.newBuilder();
+    parseJsonResource("/" + ARTIFACT_MANIFEST_PATH, builder);
+    return builder.build();
+  }
+
+  /** Writes artifacts listed in {@code proxyManifest}. */
+  public static String stageArtifacts(
+      ProxyManifest proxyManifest,
+      PipelineOptions options,
+      String invocationId,
+      String artifactStagingPath)
+      throws Exception {
+    Collection<StagedFile> filesToStage =
+        prepareArtifactsForStaging(proxyManifest, options, invocationId);
+    try (GrpcFnServer artifactServer =
+        GrpcFnServer.allocatePortAndCreateFor(
+            new BeamFileSystemArtifactStagingService(), InProcessServerFactory.create())) {
+      ManagedChannel grpcChannel =
+          InProcessManagedChannelFactory.create()
+              .forDescriptor(artifactServer.getApiServiceDescriptor());
+      ArtifactServiceStager stager = ArtifactServiceStager.overChannel(grpcChannel);
+      String stagingSessionToken =
+          BeamFileSystemArtifactStagingService.generateStagingSessionToken(
+              invocationId, artifactStagingPath);
+      String retrievalToken = stager.stage(stagingSessionToken, filesToStage);
+      // Clean up.
+      for (StagedFile file : filesToStage) {
+        if (!file.getFile().delete()) {
+          LOG.warn("Failed to delete file {}", file.getFile());
+        }
+      }
+      grpcChannel.shutdown();
+      return retrievalToken;
+    }
+  }
+
+  /**
+   * Artifacts are expected to exist as resources on the classpath, located using {@code
+   * proxyManifest}. Write them to tmp files so they can be staged.
+   */
+  private static Collection<StagedFile> prepareArtifactsForStaging(
+      ProxyManifest proxyManifest, PipelineOptions options, String invocationId)
+      throws IOException {
+    List<StagedFile> filesToStage = new ArrayList<>();
+    Path outputFolderPath =
+        Paths.get(
+            MoreObjects.firstNonNull(
+                options.getTempLocation(), System.getProperty("java.io.tmpdir")),
+            invocationId);
+    if (!outputFolderPath.toFile().mkdir()) {
+      throw new IOException("Failed to create folder " + outputFolderPath);
+    }
+    for (Location location : proxyManifest.getLocationList()) {
+      try (InputStream inputStream = getResourceFromClassPath(location.getUri())) {
```

Comment:
Can we use the jar directly instead of copying the artifacts to local file system?
Copying files to local file system will also call for cleanup.

Reply:
As far as I know, the jar's contents only exist as class path resources (streams), not regular files, and the artifact staging code expects that we are staging regular files.